hooks/provider-pinned-versions/required_providers.awk

```diff
@@ -0,0 +1,54 @@
+#! /bin/awk
+
+BEGIN {
+    in_required_providers = 0;
+    brace_count = 0;
+    version_prefix_regex = ".*version[[:space:]]+=[[:space:]]+";
+    provider_prefix_regex = "[a-z_-]+[[:space:]]+=[[:space:]]+\{";
+    pinned_version_regex = "[0-9]+\.[0-9]+\.[0-9]+";
+    version_constraints_regex = "[!~><]+";
+}
+
+{
+    if ($0 ~ /required_providers/) {
+        in_required_providers = 1;
+    }
+
+    # If inside "required_providers" block, count braces
+    if (in_required_providers) {
+        brace_count += gsub(/{/, "{");
+        brace_count -= gsub(/}/, "}");
+
+        # If brace count returns to 0, exit the block
+        if (brace_count == 0) {
+            in_required_providers = 0;
+            next;
+        }
+
+        if ($0 ~ provider_prefix_regex) {
+            provider = $1;
+        }
+
+        # If the line doesn't contain a version, skip
+        if (!($0 ~ version_prefix_regex)) {
+            next;
+        } else {
+            # If the line contains a version, remove the prefix
+            gsub(version_prefix_regex, "", $0);
```

Comment:
I would have used [match](https://www.gnu.org/software/gawk/manual/html_node/String-Functions.html#index-match_0028_0029-function) with capture groups but it wasn't working on macos.  I think it's a gawk specific function and depends on the os.